site/src/lib/model/apiKey.model.ts

```diff
@@ -0,0 +1,239 @@
+
+    return `${this.PREFIX}.${publicId}.${privateId}`;
+  }
+
+  private static async getByPublicId(
+    db: Db,
+    params: { publicId: string },
+  ): Promise<ApiKey | undefined> {
+    const apiKeyRecord = await db
+      .collection<ApiKeyDocument>(this.COLLECTION_NAME)
+      .findOne({
+        publicId: params.publicId,
+      });
+
+    if (!apiKeyRecord) {
+      return undefined;
+    }
+
+    return new ApiKey(apiKeyRecord);
+  }
+
+  static async getByUser(
+    db: Db,
+    params: { user: User },
+  ): Promise<UserFacingApiKeyProperties[]> {
+    const projection: Record<keyof UserFacingApiKeyProperties, 1> = {
+      createdAt: 1,
+      displayName: 1,
+      lastUsedAt: 1,
+      publicId: 1,
+      revokedAt: 1,
+      useCount: 1,
+    };
+
+    return await db
+      .collection<UserFacingApiKeyProperties>(this.COLLECTION_NAME)
+      .find(
+        { user: params.user.toRef() },
+        {
+          projection: {
+            ...projection,
+            _id: 0,
+          },
+        },
+      )
+      .toArray();
+  }
+
+  static async validate(
+    db: Db,
+    params: { apiKeyString: string },
+  ): Promise<boolean> {
+    if (!ApiKey.HASHING_SECRET) {
+      throw new Error(
+        "Cannot validate API key: no HASHING_SECRET present in environment.",
+      );
+    }
+
+    const { apiKeyString } = params;
+
+    if (!apiKeyString.match(this.KEY_FORMAT_REGEXP)) {
+      throw new Error("API key does not match the expected format.");
+    }
+
+    const INVALID_KEY_ERROR_MSG = "Invalid API key.";
+
+    const [_prefix, publicId, privateId] = apiKeyString.split(".");
+
+    const apiKey = await ApiKey.getByPublicId(db, { publicId });
+
+    if (!apiKey) {
+      throw new Error(INVALID_KEY_ERROR_MSG);
+    }
+
+    const providedKeyHash = ApiKey.hash({
+      publicId,
+      privateId,
+      salt: apiKey.salt,
+    });
+    if (providedKeyHash !== apiKey.hashedString) {
+      throw new Error(INVALID_KEY_ERROR_MSG);
+    } else if (apiKey.isRevoked()) {
+      throw new Error("API key has been revoked.");
+    }
+
+    await apiKey.registerUse(db);
+
+    return true;
+  }
+
+  static async revokeAll(db: Db, params: { user: User }): Promise<void> {
+    await db
+      .collection<ApiKeyDocument>(ApiKey.COLLECTION_NAME)
+      .updateMany(
+        { user: params.user.toRef(), revokedAt: { $eq: null } },
+        { $set: { revokedAt: new Date() } },
+      );
+  }
+
+  isRevoked() {
+    return this.revokedAt && this.revokedAt.valueOf() <= new Date().valueOf();
+  }
+
+  async registerUse(db: Db) {
+    const { value } = await db
+      .collection<ApiKeyDocument>(ApiKey.COLLECTION_NAME)
+      .findOneAndUpdate(
+        { publicId: this.publicId },
+        { $inc: { useCount: 1 }, $set: { lastUsedAt: new Date() } },
+        { returnDocument: "after" },
+      );
+
+    if (value) {
+      this.lastUsedAt = value.lastUsedAt;
+      this.useCount = value.useCount;
+    }
+  }
```

Comment:
What do we need `useCount` for? Doing too many writes may become a bottleneck 🤔 If we only update `lastUsedAt`, we can do this with some threshold (like at most once per minute).

Reply:
Some indication of activity would be helpful - until we have better event tracking generally, and levels of activity that cause concern, I think this is fine.